File: app/src/main/java/com/trevorwiebe/trackacow/presentation/activities/MarkACowDeadActivity.java
package com.trevorwiebe.trackacow.presentation.activities;

import android.app.DatePickerDialog;

import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;

import androidx.cardview.widget.CardView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.TextView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.domain.dataLoaders.cache.holdingCow.InsertHoldingCow;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.cow.InsertSingleCow;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.cow.QueryDeadCowsByLotIds;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.lot.QueryLotsByPenId;
import com.trevorwiebe.trackacow.data.db.entities.CowEntity;
import com.trevorwiebe.trackacow.data.db.entities.LotEntity;
import com.trevorwiebe.trackacow.data.db.holdingUpdateEntities.HoldingCowEntity;
import com.trevorwiebe.trackacow.domain.utils.Constants;
import com.trevorwiebe.trackacow.domain.utils.Utility;

import java.util.ArrayList;
import java.util.Calendar;

public class MarkACowDeadActivity extends AppCompatActivity implements
        QueryDeadCowsByLotIds.OnDeadCowsLoaded,
        QueryLotsByPenId.OnLotsByPenIdLoaded {

    private CardView mDeadCowCard;
    private TextInputEditText mTagNumber;
    private TextInputEditText mDate;
    private TextInputEditText mNotes;
    private TextInputLayout mAddNotesLayout;
    private Button mAddNotes;

    private DatePickerDialog.OnDateSetListener mStartDatePicker;
    private Calendar mCalendar = Calendar.getInstance();
    private LotEntity mSelectedLot;
    private DatabaseReference mBaseRef;
    private ArrayList<CowEntity> mDeadCowList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_mark_a_cow_dead);

        this.getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close_white_24dp);
        this.getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        String penId = getIntent().getStringExtra("penId");

        new QueryLotsByPenId(penId, this).execute(this);

        mBaseRef = FirebaseDatabase.getInstance().getReference("users").child(FirebaseAuth.getInstance().getCurrentUser().getUid());

        mDeadCowCard = findViewById(R.id.cow_is_dead_already_card);
        mTagNumber = findViewById(R.id.deads_tag_number);
        mDate = findViewById(R.id.deads_date);
        mNotes = findViewById(R.id.deads_notes);
        mAddNotesLayout = findViewById(R.id.dead_notes_layout);
        mAddNotes = findViewById(R.id.dead_add_notes_btn);

        mTagNumber.setOnEditorActionListener(doneListener);
        mNotes.setOnEditorActionListener(doneListener);

        mDate.setText(Utility.convertMillisToDate(mCalendar.getTimeInMillis()));
        mDate.setOnClickListener(new android.view.View.OnClickListener() {
            @Override
            public void onClick(android.view.View view) {
                new DatePickerDialog(MarkACowDeadActivity.this,
                        mStartDatePicker,
                        mCalendar.get(Calendar.YEAR),
                        mCalendar.get(Calendar.MONTH),
                        mCalendar.get(Calendar.DAY_OF_MONTH))
                        .show();
            }
        });
        mStartDatePicker = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                mCalendar.set(Calendar.YEAR, year);
                mCalendar.set(Calendar.MONTH, month);
                mCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                mDate.setText(Utility.convertMillisToDate(mCalendar.getTimeInMillis()));
            }
        };

        mTagNumber.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() != 0) {
                    int tagNumber = Integer.parseInt(s.toString());
                    CowEntity cowEntity = findCowEntity(tagNumber);
                    if (cowEntity != null) {
                        mDeadCowCard.setVisibility(View.VISIBLE);
                    } else {
                        mDeadCowCard.setVisibility(View.GONE);
                    }
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        mAddNotes.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mAddNotesLayout.setVisibility(View.VISIBLE);
                mAddNotes.setVisibility(View.GONE);
            }
        });
    }

    TextView.OnEditorActionListener doneListener = new TextView.OnEditorActionListener() {
        @Override
        public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
            if ((event != null && (event.getKeyCode() == KeyEvent.KEYCODE_ENTER)) || (actionId == EditorInfo.IME_ACTION_DONE)) {
                markAsDead(null);
            }
            return false;
        }
    };

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return false;
    }

    public void markAsDead(View view) {
        if(mTagNumber.length() == 0 || mDate.length() == 0 ){
            mTagNumber.requestFocus();
            mTagNumber.setError("Please fill the blank");
            return;
        }

        int tagNumber = Integer.parseInt(mTagNumber.getText().toString());
        String notes = mNotes.getText().toString();

        DatabaseReference pushRef = mBaseRef.child(CowEntity.COW).push();

        CowEntity cowEntity = new CowEntity(0, pushRef.getKey(), tagNumber, mCalendar.getTimeInMillis(), notes, mSelectedLot.getLotId());

        if(Utility.haveNetworkConnection(this)){
            pushRef.setValue(cowEntity);
        }else{
            Utility.setNewDataToUpload(MarkACowDeadActivity.this, true);
            HoldingCowEntity holdingCowEntity = new HoldingCowEntity(cowEntity, Constants.INSERT_UPDATE);
            new InsertHoldingCow(holdingCowEntity).execute(MarkACowDeadActivity.this);
        }

        new InsertSingleCow(cowEntity).execute(this);

        mDeadCowList.add(cowEntity);

        mAddNotes.setVisibility(View.VISIBLE);
        mAddNotesLayout.setVisibility(View.GONE);

        mTagNumber.setText("");
        mNotes.setText("");
        mCalendar = Calendar.getInstance();
        mDate.setText(Utility.convertMillisToDate(mCalendar.getTimeInMillis()));
        mTagNumber.requestFocus();

        InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY);
    }

    @Override
    public void onDeadCowsLoaded(ArrayList<CowEntity> cowEntities) {
        mDeadCowList = cowEntities;
    }

    private CowEntity findCowEntity(int tagNumber) {
        for (int v = 0; v < mDeadCowList.size(); v++) {
            CowEntity cowEntity = mDeadCowList.get(v);
            if (cowEntity.getTagNumber() == tagNumber) {
                return cowEntity;
            }
        }
        return null;
    }

    @Override
    public void onLotsByPenIdLoaded(ArrayList<LotEntity> lotEntities) {
        mSelectedLot = lotEntities.get(0);
        ArrayList<String> lotIds = new ArrayList<>();
        for (int e = 0; e < lotEntities.size(); e++) {
            LotEntity lotEntity = lotEntities.get(e);
            String lotId = lotEntity.getLotId();
            lotIds.add(lotId);
        }

        new QueryDeadCowsByLotIds(MarkACowDeadActivity.this, lotIds).execute(MarkACowDeadActivity.this);
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/mark_a_cow_dead/MarkACowDeadActivity.kt
package com.trevorwiebe.trackacow.presentation.mark_a_cow_dead

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.Button
import android.widget.TextView.OnEditorActionListener
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.trevorwiebe.trackacow.R
import com.trevorwiebe.trackacow.domain.models.compound_model.PenAndLotModel
import com.trevorwiebe.trackacow.domain.models.cow.CowModel
import com.trevorwiebe.trackacow.domain.utils.Utility
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import java.util.*
import javax.inject.Inject

@AndroidEntryPoint
class MarkACowDeadActivity : AppCompatActivity() {

    private lateinit var mDeadCowCard: CardView
    private lateinit var mTagNumber: TextInputEditText
    private lateinit var mDate: TextInputEditText
    private lateinit var mNotes: TextInputEditText
    private lateinit var mAddNotesLayout: TextInputLayout
    private lateinit var mAddNotes: Button
    private lateinit var mStartDatePicker: OnDateSetListener

    private var mCalendar = Calendar.getInstance()
    private var mDeadCowList: MutableList<CowModel> = mutableListOf()
    private var mPenAndLotModel: PenAndLotModel? = null

    @Inject
    lateinit var markACowDeadViewModelFactory: MarkACowDeadViewModel.MarkACowDeadViewModelFactory

    @Suppress("DEPRECATION")
    private val markACowDeadViewModel: MarkACowDeadViewModel by viewModels {
        MarkACowDeadViewModel.providesFactory(
            assistedFactory = markACowDeadViewModelFactory,
            lotId = if (Build.VERSION.SDK_INT >= 33) {
                intent.getParcelableExtra("penAndLotModel", PenAndLotModel::class.java)
            } else {
                intent.getParcelableExtra("penAndLotModel")
            }
                ?.lotCloudDatabaseId ?: ""
        )
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_mark_a_cow_dead)

        this.supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)
        this.supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        @Suppress("DEPRECATION")
        mPenAndLotModel = if (Build.VERSION.SDK_INT >= 33) {
            intent.getParcelableExtra("penAndLotModel", PenAndLotModel::class.java)
        } else {
            intent.getParcelableExtra("penAndLotModel")
        }

        mDeadCowCard = findViewById(R.id.cow_is_dead_already_card)
        mTagNumber = findViewById(R.id.deads_tag_number)
        mDate = findViewById(R.id.deads_date)
        mNotes = findViewById(R.id.deads_notes)
        mAddNotesLayout = findViewById(R.id.dead_notes_layout)
        mAddNotes = findViewById(R.id.dead_add_notes_btn)
        mTagNumber.setOnEditorActionListener(doneListener)
        mNotes.setOnEditorActionListener(doneListener)
        mDate.setText(Utility.convertMillisToDate(mCalendar.timeInMillis))
        mDate.setOnClickListener {
            DatePickerDialog(
                this@MarkACowDeadActivity,
                mStartDatePicker,
                mCalendar[Calendar.YEAR],
                mCalendar[Calendar.MONTH],
                mCalendar[Calendar.DAY_OF_MONTH]
            )
                .show()
        }

        mStartDatePicker = OnDateSetListener { _, year, month, dayOfMonth ->
            mCalendar[Calendar.YEAR] = year
            mCalendar[Calendar.MONTH] = month
            mCalendar[Calendar.DAY_OF_MONTH] = dayOfMonth
            mDate.setText(Utility.convertMillisToDate(mCalendar.timeInMillis))
        }

        mTagNumber.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (s.isNotEmpty()) {


                    if (mDeadCowList.any { it.tagNumber == s.toString().toInt() }) {
                        mDeadCowCard.visibility = View.VISIBLE
                    } else {
                        mDeadCowCard.visibility = View.GONE
                    }
                }
            }

            override fun afterTextChanged(s: Editable) {}
        })

        mAddNotes.setOnClickListener {
            mAddNotesLayout.visibility = View.VISIBLE
            mAddNotes.visibility = View.GONE
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                markACowDeadViewModel.uiState.collect {
                    mDeadCowList = it.deadCowList.toMutableList()
                }
            }
        }

    }

    private var doneListener = OnEditorActionListener { v, actionId, event ->
        if (event != null && event.keyCode == KeyEvent.KEYCODE_ENTER || actionId == EditorInfo.IME_ACTION_DONE) {
            markAsDead(null)
        }
        false
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return false
    }

    fun markAsDead(view: View?) {
        if (mTagNumber.length() == 0 || mDate.length() == 0) {
            mTagNumber.requestFocus()
            mTagNumber.error = "Please fill the blank"
            return
        }
        val tagNumber = mTagNumber.text.toString().toInt()
        val notes = mNotes.text.toString()

        val cowModel = CowModel(
            0,
            0,
            "",
            tagNumber,
            mCalendar.timeInMillis,
            notes,
            mPenAndLotModel?.lotCloudDatabaseId ?: ""
        )

        markACowDeadViewModel.onEvent(MarkACowDeadEvent.OnDeadCowCreated(cowModel))

        mDeadCowList.add(cowModel)

        mAddNotes.visibility = View.VISIBLE

        mDeadCowCard.visibility = View.GONE

        mAddNotesLayout.visibility = View.GONE
        mTagNumber.setText("")
        mNotes.setText("")
        mCalendar = Calendar.getInstance()
        mDate.setText(Utility.convertMillisToDate(mCalendar.timeInMillis))
        mTagNumber.requestFocus()
    }
}